File: android/app/src/main/java/com/noop/push/ServerHrvSeries.kt
package com.noop.push

import org.json.JSONObject

/** Selected server RMSSD windows only; never substitutes a daily scalar or SDNN. */
data class ServerHrvSeries(
    val windows: List<Window>, val featureStatus: String?, val featureReason: String?,
    val deviceId: String?, val algorithmVersion: String?, val observedThrough: String?, val stale: Boolean,
) {
    data class Window(
        val start: Long, val end: Long, val rmssdMs: Double?, val measurementValid: Boolean,
        val context: String, val baselineEligible: Boolean, val reason: String?, val baselineReason: String?,
        val baselineEffectiveSampleCount: Int?, val baselineRobustZ: Double?, val source: String?,
        val modality: String?, val methodVersion: String?, val observedTimeFraction: Double?, val isGap: Boolean,
    )
    companion object {
        fun from(cache: ServerScoreDayCache?, day: String): ServerHrvSeries {
            val feature = cache?.features?.get("hrv")
            fun result(rows: List<Window>) = ServerHrvSeries(rows, feature?.status, feature?.reason, feature?.deviceId,
                feature?.algorithmVersion, feature?.observedThrough, cache?.stale ?: true)
            if (cache == null || cache.day != day || cache.ownerId.isBlank() || cache.schemaVersion != 2 || feature == null ||
                feature.algorithmVersion != "frwhoop-physiology-2" || feature.deviceId.isNullOrEmpty() || feature.inputRevision == null) {
                return result(emptyList())
            }
            val overlay = runCatching { JSONObject(cache.rawSnapshotJSON ?: "{}").optJSONObject("server_scoring") }.getOrNull()
                ?: return result(emptyList())
            val selected = overlay.optJSONObject("features")?.optJSONObject("hrv") ?: return result(emptyList())
            if (overlay.optString("user_id").lowercase() != cache.ownerId.lowercase() || overlay.optString("day") != day ||
                selected.optString("device_id") != feature.deviceId || selected.optString("algorithm_version") != feature.algorithmVersion ||
                selected.opt("input_revision")?.toString() != feature.inputRevision.toString()) return result(emptyList())
            val rows = overlay.optJSONArray("measurements") ?: return result(emptyList())
            fun JSONObject.text(key: String) = opt(key) as? String
            fun JSONObject.number(key: String) = (opt(key) as? Number)?.toDouble()?.takeIf { it.isFinite() }
            fun JSONObject.integer(key: String) = number(key)?.takeIf {
                it == kotlin.math.floor(it) && it >= -62135596800 && it <= 253402300799
            }?.toLong()
            val parsed = (0 until rows.length()).mapNotNull { index ->
                val row = rows.optJSONObject(index) ?: return@mapNotNull null
                val start = row.integer("start")
                val end = row.integer("end")
                if (row.integer("measurement_schema_version") != 1L || row.text("feature") != "hrv" ||
                    row.text("user_id")?.lowercase() != cache.ownerId.lowercase() || row.text("device_id") != feature.deviceId ||
                    row.text("metric") != "rmssd" || row.text("unit") != "ms" || start == null || end == null ||
                    start % 300 != 0L || end - start != 300L) return@mapNotNull null
                if (row.opt("input_revision")?.toString() != feature.inputRevision.toString()) return@mapNotNull null
                val value = row.number("observed_rmssd_ms")
                val source = row.text("source")
                val modality = row.text("modality")
                val valid = row.opt("measurement_valid") == true && value != null && value >= 0 && row.isNull("reason") &&
                    !source.isNullOrEmpty() && !modality.isNullOrEmpty()
                val context = row.text("context") ?: "unknown"
                val eligible = valid && row.opt("baseline_eligible") == true && context in listOf("sleep", "nap", "quiet_rest")
                val baseline = row.optJSONObject("baseline")
                Window(start, end, value.takeIf { valid }, valid, context, eligible,
                    if (valid) null else row.text("reason") ?: "measurement_unavailable",
                    row.text("baseline_reason") ?: baseline?.text("reason"),
                    baseline?.integer("effective_sample_count")?.takeIf { it in 0..Int.MAX_VALUE }?.toInt(),
                    baseline?.number("robust_z").takeIf { eligible }, source, modality, row.text("algorithm_version"),
                    row.number("observed_time_fraction")?.takeIf { it in 0.0..1.0 }, false)
            }
            val buckets = parsed.groupBy { it.start }
            val first = buckets.keys.minOrNull()
            val last = buckets.keys.maxOrNull()
            if (first == null || last == null || last - first > 76 * 3600) return result(emptyList())
            return result((first..last step 300).map { start ->
                val values = buckets[start]
                if (values == null) gap(start, "window_missing") else if (values.distinct().size == 1) values.first()
                else gap(start, "conflicting_window_records")
            })
        }
        private fun gap(start: Long, reason: String) = Window(start, start + 300, null, false, "unknown", false, reason,
            "measurement_unavailable", null, null, null, null, null, null, true)
    }
}

File: Packages/WhoopStore/Sources/WhoopStore/ServerHrvSeries.swift
import Foundation
import CoreFoundation

/// Display projection of the selected server RMSSD series; never substitutes a daily scalar or SDNN.
public struct ServerHrvSeries {
    public struct Window: Equatable, Identifiable {
        public var id: Int { start }
        public let start: Int, end: Int
        public let rmssdMs: Double?
        public let measurementValid: Bool
        public let context: String
        public let baselineEligible: Bool
        public let reason: String?, baselineReason: String?
        public let baselineEffectiveSampleCount: Int?
        public let baselineRobustZ: Double?
        public let source: String?, modality: String?, methodVersion: String?
        public let observedTimeFraction: Double?
        public let isGap: Bool
    }
    public let windows: [Window]
    public let featureStatus: String?, featureReason: String?
    public let deviceId: String?, algorithmVersion: String?, observedThrough: String?
    public let stale: Bool

    public static func from(_ cache: ServerScoreDayCache?, day: String) -> Self {
        let feature = cache?.features["hrv"]
        func result(_ windows: [Window]) -> Self {
            Self(windows: windows, featureStatus: feature?.status, featureReason: feature?.reason,
                deviceId: feature?.deviceId, algorithmVersion: feature?.algorithmVersion,
                observedThrough: feature?.observedThrough, stale: cache?.stale ?? true)
        }
        guard let cache, cache.day == day, !cache.ownerId.isEmpty, cache.schemaVersion == 2,
              let feature, feature.algorithmVersion == "frwhoop-physiology-2", let device = feature.deviceId, !device.isEmpty,
              let revision = feature.inputRevision,
              let json = cache.rawSnapshotJSON?.data(using: .utf8),
              let root = (try? JSONSerialization.jsonObject(with: json)) as? [String: Any],
              let overlay = root["server_scoring"] as? [String: Any],
              (overlay["user_id"] as? String)?.lowercased() == cache.ownerId.lowercased(), overlay["day"] as? String == day,
              let selected = (overlay["features"] as? [String: [String: Any]])?["hrv"],
              selected["device_id"] as? String == device, selected["algorithm_version"] as? String == feature.algorithmVersion,
              (selected["input_revision"] as? NSNumber)?.int64Value == revision,
              let rows = overlay["measurements"] as? [Any] else { return result([]) }
        func number(_ value: Any?) -> Double? {
            guard let value = value as? NSNumber, CFGetTypeID(value) != CFBooleanGetTypeID(), value.doubleValue.isFinite else { return nil }
            return value.doubleValue
        }
        func integer(_ value: Any?) -> Int? {
            guard let n = number(value), n.rounded() == n, n >= -62135596800, n <= 253402300799 else { return nil }
            return Int(n)
        }
        let parsed = rows.compactMap { raw -> Window? in
            guard let row = raw as? [String: Any] else { return nil }
            guard integer(row["measurement_schema_version"]) == 1, row["feature"] as? String == "hrv",
                  (row["user_id"] as? String)?.lowercased() == cache.ownerId.lowercased(), row["device_id"] as? String == device,
                  row["metric"] as? String == "rmssd", row["unit"] as? String == "ms",
                  let start = integer(row["start"]), let end = integer(row["end"]), start % 300 == 0, end - start == 300 else { return nil }
            let rawRevision = (row["input_revision"] as? String) ?? (row["input_revision"] as? NSNumber)?.stringValue
            guard rawRevision == String(revision) else { return nil }
            let value = number(row["observed_rmssd_ms"])
            let source = row["source"] as? String, modality = row["modality"] as? String
            let valid = (row["measurement_valid"] as? NSNumber).map { CFGetTypeID($0) == CFBooleanGetTypeID() && $0.boolValue } == true &&
                value.map { $0 >= 0 } == true && (row["reason"] == nil || row["reason"] is NSNull) &&
                source?.isEmpty == false && modality?.isEmpty == false
            let context = row["context"] as? String ?? "unknown"
            let eligible = valid && row["baseline_eligible"] as? Bool == true && ["sleep", "nap", "quiet_rest"].contains(context)
            let baseline = row["baseline"] as? [String: Any]
            let coverage = number(row["observed_time_fraction"]).flatMap { (0...1).contains($0) ? $0 : nil }
            return Window(start: start, end: end, rmssdMs: valid ? value : nil, measurementValid: valid,
                context: context, baselineEligible: eligible,
                reason: valid ? nil : (row["reason"] as? String ?? "measurement_unavailable"),
                baselineReason: (row["baseline_reason"] as? String) ?? (baseline?["reason"] as? String),
                baselineEffectiveSampleCount: integer(baseline?["effective_sample_count"]).flatMap { $0 >= 0 ? $0 : nil },
                baselineRobustZ: eligible ? number(baseline?["robust_z"]) : nil,
                source: source, modality: modality, methodVersion: row["algorithm_version"] as? String,
                observedTimeFraction: coverage, isGap: false)
        }
        let buckets = Dictionary(grouping: parsed, by: \.start)
        guard let first = buckets.keys.min(), let last = buckets.keys.max(), last - first <= 76 * 3600 else { return result([]) }
        let windows = stride(from: first, through: last, by: 300).map { start -> Window in
            guard let values = buckets[start], let row = values.first else { return gap(start, "window_missing") }
            return values.allSatisfy { $0 == row } ? row : gap(start, "conflicting_window_records")
        }
        return result(windows)
    }

    private static func gap(_ start: Int, _ reason: String) -> Window {
        Window(start: start, end: start + 300, rmssdMs: nil, measurementValid: false, context: "unknown",
            baselineEligible: false, reason: reason, baselineReason: "measurement_unavailable",
            baselineEffectiveSampleCount: nil, baselineRobustZ: nil, source: nil, modality: nil,
            methodVersion: nil, observedTimeFraction: nil, isGap: true)
    }
}
